### PR TITLE
doc: improve binding tutorial

### PR DESCRIPTION
**Pull request by raphi011**
## Basics

Check relevant points but **please do not remove entries**.
Do not describe the purpose of this PR in the PR description but:

- [ ] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [ ] Longer descriptions should be in documentation or in design decisions.
- [ ] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise and good to read
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the build server is happy** and also you
  say that everything is ready to be merged.

Issue  #2928

### CODE REVIEW

**Comment by markus2330** [comment]
Thank you for creating this PR!

You can also extend it in the same PR.

**Comment by raphi011** [comment]
@markus2330, I've added some of my experiences - take a look

File: doc/tutorials/highlevel-bindings.md [hunk 1]
```diff
@@ -13,17 +13,17 @@ The goals of any high-level API (or binding to the C high-level API) for Elektra
...
-  - **No Errors in Getters:** It is a stated goal of our high-level API to make `get` calls not able to fail. This means
-    `get` calls _cannot_ return an error under normal conditions. This can be ensure via checks during initialization or
-    via code-generation. Both are based on the specification. Without a specification it is not possible to prevent
-    errors because of missing keys, since we have no way of knowing which keys should exist. <br/><br/>
-    There is an exception to this rule. Some target languages have a standard error concept, which not only forces the
-    user to handle arising errors, but also does so in a simple and concise way. An example of this is Rust. Its `Result`
-    type forces the user to handle errors and the `?` operator allows to do this in a concise way. <br/>
-    However, we still recommend to avoid errors as far as possible.
-- **Idomatic Code:** If the targt language has a concept of "idomatic code", the API should fall into that category.
+- **No Errors in Getters:** It is a stated goal of our high-level API to make `get` calls not able to fail. This means
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]: Did you remove this space on purpose?

File: doc/tutorials/highlevel-bindings.md [hunk 2]
```diff
@@ -54,6 +54,29 @@ sufficient to write (or generate) a one-to-one mirror of the C API and use that
...
 direct mapping, significantly simplifies the generated code (and maybe also the template), you should write such an API.
 
+### Bindings for the `lowlevel` Library
+
+If you decide on mapping the functionality of kdb.h 1:1 it is pretty straight-forward.
+
+Your programming language of choice must provide a way to call into C code, e.g.: 
+* [cgo](https://golang.org/cmd/cgo/) for the Go programming language
+* [jni](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/) for JAVA
+* [n-api](https://nodejs.org/api/n-api.html) for javascript
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]: I do not think these examples are relevant. And n-api is a bad example.

File: doc/tutorials/highlevel-bindings.md [hunk 3]
```diff
@@ -54,6 +54,29 @@ sufficient to write (or generate) a one-to-one mirror of the C API and use that
...
+
+Make sure you have a good understanding of these libraries because some of them can have unexpected downsides.
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]: About which libraries do you talk here? You mean that there might be different ways of do bindings?

In general we prefer (in this order):

- automatically and statically generated bindings (like swig)
- manually written bindings that directly call to library without any statically build part (like JNI and GI). They are a bit slower but can directly access libelektra.so.
- manually written bindings that are built statically

File: doc/tutorials/highlevel-bindings.md [hunk 4]
```diff
@@ -54,6 +54,29 @@ sufficient to write (or generate) a one-to-one mirror of the C API and use that
...
+CGO for example cannot call variadic C functions because in C the amount of parameters has to be known at compile-time. This is not the case with Go
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]: Give variadic C functions an own section as this is always something problematic.

@PhilippGackstatter how did you solve it?

File: doc/tutorials/highlevel-bindings.md [hunk 5]
```diff
@@ -54,6 +54,29 @@ sufficient to write (or generate) a one-to-one mirror of the C API and use that
...
+since it supports variable length arguments at runtime with the `...` operator. 
+
+This is unfortunate because the low-level bindings rely heavily on variadic functions. It is possible to work around this problem by writing helper functions in C that call these variadic functions with a fixed amount of parameters.
+
+What you will also need is to set up the Compiler + Linker flags, for this I can recommend [pkg-config](https://www.freedesktop.org/wiki/Software/pkg-config/), because Elektra already provides `.pc` files.
+
+For garbage collected (GC) languages freeing memory by hand is not something you usually do and since the GC has no knowledge of memory allocated in C we have two options:
+* Forcing the user to call the appropriate functions like `keyDel()` themselves, which is very developer unfriendly and error prone or
+* Using language featurse like Java / C++ Destructors or Go's `runtime.SetFinalizer()` function to automatically and reliably release the memory as soon as the native objects are garbage collected
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]:
```suggestion
* Using language featurse like Java / C++ Destructors or Go's `runtime.SetFinalizer()` function to automatically and reliably release the memory as soon as the native objects are garbage collected.
```

File: doc/tutorials/highlevel-bindings.md [hunk 6]
```diff
@@ -54,6 +54,29 @@ sufficient to write (or generate) a one-to-one mirror of the C API and use that
...
+
+If you want to adapt or enhance some API's to leverage language features like iterators or operator overloading feel free to do so. The less 'alien' the binding feels to its users the better.
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]: Very vague. Important for iterators is that Elektra has internal iterators (for metadata+keysets) but in general we prefer external iterators. (by either copying the KeySet per iterator or using ksAtCursor).

File: doc/tutorials/highlevel-bindings.md (hunk 1 above)
Comment (raphi011) [reply doc/tutorials/highlevel-bindings.md]: Yes - it seemed to me as if it was not intended - if it was i can undo this change

File: doc/tutorials/highlevel-bindings.md (hunk 3 above)
Comment (raphi011) [reply doc/tutorials/highlevel-bindings.md]: Maybe library is the wrong word - I've meant the 'interop' layer to communicate with C, like cgo.

> In general we prefer (in this order):
>* automatically and statically generated bindings (like swig)
>* manually written bindings that directly call to library without any statically build part (like JNI and GI). They are a bit slower but can directly access libelektra.so.
>* manually written bindings that are built statically

Do you want me to add this?

File: doc/tutorials/highlevel-bindings.md (hunk 2 above)
Comment (raphi011) [reply doc/tutorials/highlevel-bindings.md]: I will remove them

File: doc/tutorials/highlevel-bindings.md (hunk 6 above)
Comment (raphi011) [reply doc/tutorials/highlevel-bindings.md]: How do you think I could improve this part?

File: doc/tutorials/highlevel-bindings.md (hunk 4 above)
Comment (PhilippGackstatter) [reply doc/tutorials/highlevel-bindings.md]: Rust allows calling variadic functions, but not declaring them. I described this as a possible use case for the raw bindings in Rust [here](https://github.com/ElektraInitiative/libelektra/tree/master/src/bindings/rust#raw-bindings).
A more general solution, applicable to many languages is to us the builder pattern. With this, it's easy to set multiple meta values in one go, just like with the variadic `keyNew`. For example, in the Rust binding.

```rust
let key: StringKey = KeyBuilder::new("user/test/key")?
            .value("value")
            .meta("metaname", "metavalue")?
            .meta("OWNER", "me")?
            .build();
```

File: doc/tutorials/highlevel-bindings.md (hunk 6 above)
Comment (markus2330) [reply doc/tutorials/highlevel-bindings.md]: I already wrote what is important to write about.

File: doc/tutorials/highlevel-bindings.md (hunk 3 above)
Comment (markus2330) [reply doc/tutorials/highlevel-bindings.md]: > Maybe library is the wrong word - I've meant the 'interop' layer to communicate with C, like cgo.

Ok.

> Do you want me to add this?

Yes, you can add it.

File: doc/tutorials/highlevel-bindings.md (hunk 1 above)
Comment (markus2330) [reply doc/tutorials/highlevel-bindings.md]: You are right, it improves the alignment.

File: doc/tutorials/highlevel-bindings.md [hunk 7]
```diff
@@ -54,6 +54,34 @@ sufficient to write (or generate) a one-to-one mirror of the C API and use that
...
+### Bindings for the `lowlevel` Library
+
+Your programming language of choice must provide a way to call into C code (like [cgo](https://golang.org/cmd/cgo/)).
+
+In general we prefer (in this order):
+1. Automatically and statically generated bindings (like [swig](http://www.swig.org/)).
+1. Manually written bindings that directly call to library without any statically build part (like JNI and GI). They are a bit slower but can directly access libelektra.so.
+1. Manually written bindings that are built statically.
+
+If you want to manually write a binding make sure you have a good understanding of the possible limitations the interop layer can have (e.g. variadic functions).
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]: I think this is generally valid and not only a problem for manually written bindings.

Which other problems might there be next to the varadic functions?

File: doc/tutorials/highlevel-bindings.md [hunk 8]
```diff
@@ -54,6 +54,34 @@ sufficient to write (or generate) a one-to-one mirror of the C API and use that
...
+
+What you will also need is to set up the Compiler + Linker flags, for this I can recommend [pkg-config](https://www.freedesktop.org/wiki/Software/pkg-config/), because Elektra already provides `.pc` files.
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]: Elektra also provides CMake files.

File: doc/tutorials/highlevel-bindings.md [hunk 9]
```diff
@@ -54,6 +54,34 @@ sufficient to write (or generate) a one-to-one mirror of the C API and use that
...
+
+For garbage collected (GC) languages freeing memory by hand is not something you usually do and since the GC has no knowledge of memory allocated in C we have two options:
+* Forcing the user to call the appropriate functions like `keyDel()` themselves, which is very developer unfriendly and error prone or
+* Using language featurse like Java / C++ Destructors or Go's `runtime.SetFinalizer()` function to automatically and reliably release the memory as soon as the native objects are garbage collected.
+
+If you decide on mapping the functionality of kdb.h 1:1 it is pretty straight-forward - whereas if you want to adapt or enhance some API's to leverage language features like iterators or operator overloading feel free to do so. The less 'alien' the binding feels to its users the better.
+
+Remember that Elektra has internal iterators (for metadata+keysets) but in general we prefer external iterators by either copying the KeySet per iterator or using `ksAtCursor`.
+
+### Variadic functions
```
Comment (markus2330) [inline doc/tutorials/highlevel-bindings.md]:
```suggestion
### Variadic Functions
```

Did you format this already with cmake-format?

**Comment by markus2330** [comment]
Please fix the remaining problem and make the build server happy. Then we can merge it.

File: doc/tutorials/highlevel-bindings.md (hunk 9 above)
Comment (raphi011) [reply doc/tutorials/highlevel-bindings.md]: you mean `reformat-markdown`? Did it now

Comment (markus2330) [reply doc/tutorials/highlevel-bindings.md]: Yes, you can also run `reformat-all` to not miss any reformatting tool.

**Comment by raphi011** [comment]
Can we merge this?

**Comment by markus2330** [comment]
Thank you for improving the docu!

Please set "ready to merge" once you think it is ready to merge.

File: doc/tutorials/highlevel-bindings.md (hunk 4 above)
Comment (markus2330) [reply doc/tutorials/highlevel-bindings.md]: Actually, it would be good if we mention this in the binding tutorial.

**Pull request merged**